CRM/Member/Form/Task/PDFLetterCommon.php

```diff
@@ -64,35 +66,27 @@ public static function postProcessMembers(&$form, $membershipIDs, $skipOnHold, $
    * @param $html_message
    * @param $categories
    *
-   * @deprecated
-   *
    * @return array
+   * @throws \API_Exception
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\UnauthorizedException
+   * @deprecated
    */
   public static function generateHTML($membershipIDs, $returnProperties, $skipOnHold, $skipDeceased, $messageToken, $html_message, $categories) {
     CRM_Core_Error::deprecatedFunctionWarning('no alternative');
-    $memberships = CRM_Utils_Token::getMembershipTokenDetails($membershipIDs);
+    $memberships = Membership::get(FALSE)
+      ->addWhere('id', 'IN', $membershipIDs)
+      ->addSelect('contact_id')->execute();
     $html = [];
 
-    foreach ($membershipIDs as $membershipID) {
-      $membership = $memberships[$membershipID];
-      // get contact information
-      $contactId = $membership['contact_id'];
-      $params = ['contact_id' => $contactId];
-      //getTokenDetails is much like calling the api contact.get function - but - with some minor
-      // special handlings. It precedes the existence of the api
-      list($contacts) = CRM_Utils_Token::getTokenDetails(
-        $params,
-        $returnProperties,
-        $skipOnHold,
-        $skipDeceased,
-        NULL,
-        $messageToken,
-        'CRM_Contribution_Form_Task_PDFLetterCommon'
-      );
-
-      $tokenHtml = CRM_Utils_Token::replaceEntityTokens('membership', $membership, $html_message, $messageToken);
-      $html[] = CRM_Core_BAO_MessageTemplate::renderTemplate(['messageTemplate' => ['msg_html' => $tokenHtml], 'contactId' => $contactId])['html'];
-
+    foreach ($memberships as $membership) {
+      $html[] = CRM_Core_BAO_MessageTemplate::renderTemplate([
+        'messageTemplate' => ['msg_html' => $html_message],
+        'contactId' => $membership['contact_id'],
+        'schema' => ['contactId', 'membershipId'],
```

Comment:
@eileenmcnaughton for my own knowledge what is the purpose of schema parameter in general?

Reply:
@monishdeb yeah - I'm hoping we can flush that out in the docs - it became required (in master) to tell the token processors which entities to resolve - I'm not sure if that is good or bad. I was going to turn my attention to in once the anomalies with the actual tokens were in hand